### PR TITLE
Spike for parallel relaying between multiple channels

### PR DESCRIPTION
**Pull request by romac**
Closes: #XXX

## Description

______

For contributor use:

- [ ] Updated the __Unreleased__ section of [CHANGELOG.md](https://github.com/informalsystems/ibc-rs/blob/master/CHANGELOG.md) with the issue.
- [ ] If applicable: Unit tests written, added test to CI.
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Updated relevant documentation (`docs/`) and code comments.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.

### CODE REVIEW

File: modules/src/ics04_channel/events.rs [hunk 1]
```diff
@@ -357,11 +357,21 @@ impl CloseInit {
     pub fn port_id(&self) -> &PortId {
         &self.0.port_id
     }
-    pub fn channel_id(&self) -> &Option<ChannelId> {
-        &self.0.channel_id
+    pub fn channel_id(&self) -> &ChannelId {
+        // FIXME(romac): Rework encoding of IbcEvents which use `Attributes`
```
Comment (romac) [inline modules/src/ics04_channel/events.rs:361]: TODO: Track in issue

File: relayer-cli/src/commands/start_multi.rs [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+
+    fn worker_for_object(&mut self, object: Object, swap: bool) -> &WorkerHandle {
+        if self.workers.contains_key(&object) {
+            &self.workers[&object]
+        } else {
+            let chains = if swap {
+                self.chains.clone().swap()
+            } else {
+                self.chains.clone()
+            };
```
Comment (romac) [inline relayer-cli/src/commands/start_multi.rs:146]: TODO: Find a better way to express this.

File: relayer-cli/src/commands/start_multi.rs [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+        };
+
+        if let Err(e) = result {
+            eprintln!("worker error: {}", e);
+        }
+    }
+
+    fn run_uni_chan_path(self, path: UnidirectionalChannelPath) -> Result<(), BoxError> {
+        println!("running worker for object {:?}", path);
+        let mut link = Link::new_from_opts(
```
Comment (romac) [inline relayer-cli/src/commands/start_multi.rs:181]: TODO: whitespace

File: relayer-cli/src/commands/start_multi.rs [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+            self.chains.src.clone(),
+            self.chains.dst.clone(),
+            LinkParameters {
+                src_port_id: path.src_port_id,
+                src_channel_id: path.src_channel_id,
+            },
+        )?;
+
+        while let Ok(cmd) = self.rx.recv() {
+            link.a_to_b.relay_from_events(cmd.batch)?;
```
Comment (romac) [inline relayer-cli/src/commands/start_multi.rs:191]: TODO: Refactor `Link` and `RelayPath` to use the latter directly here.

File: relayer-cli/src/commands/start_multi.rs (hunk 2 above)
Comment (romac) [reply relayer-cli/src/commands/start_multi.rs:146]: Will address in follow-up PR for multi-chain supervisor.

File: relayer-cli/src/commands/start_multi.rs (hunk 4 above)
Comment (romac) [reply relayer-cli/src/commands/start_multi.rs:191]: Will adress in follow-up PR for multi-chain supervisor.

File: relayer/src/supervisor.rs [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+            workers: HashMap::new(),
+        })
+    }
+
+    pub fn run(mut self) -> Result<(), BoxError> {
+        let subscription_a = self.chains.a.subscribe()?;
+        let subscription_b = self.chains.b.subscribe()?;
+
+        loop {
+            println!("{} => iterating over event batches", self.chains.a.id());
```
Comment (ancazamfir) [inline relayer/src/supervisor.rs:82]: Could you remove this? And the one below?

**Review by ancazamfir — APPROVED** [review]
This looks amazing! I ran it with three different channels and all looks good. Also reviewed the code, I think we should merge this asap, can't wait to have this in master :)
Thanks Romain!

File: relayer/src/supervisor.rs (hunk 5 above)
Comment (romac) [reply relayer/src/supervisor.rs]: https://github.com/informalsystems/ibc-rs/pull/692/commits/c5f4da7f0af7773aab1da162cb81849729e743da

**Pull request merged**